core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftTransportPool.java

```diff
@@ -49,28 +50,28 @@
   private long killTime = 1000 * 3;
 
   private static class CachedConnections {
-    LinkedList<CachedConnection> unreserved = new LinkedList<>();
-    Map<CachedTTransport,CachedConnection> reserved = new HashMap<>();
+    Deque<CachedConnection> unreserved = new LinkedBlockingDeque<>(); // stack - LIFO
+    Map<CachedTTransport,CachedConnection> reserved = new ConcurrentHashMap<>();
 
     public CachedConnection reserveAny() {
-      if (unreserved.size() > 0) {
-        CachedConnection cachedConnection = unreserved.removeFirst();
+
+      CachedConnection cachedConnection = unreserved.poll(); // safe pop
+      if (cachedConnection != null) {
         cachedConnection.reserve();
         reserved.put(cachedConnection.transport, cachedConnection);
         if (log.isTraceEnabled()) {
           log.trace("Using existing connection to {}", cachedConnection.transport.cacheKey);
         }
-        return cachedConnection;
       }
-
-      return null;
+      return cachedConnection;
     }
   }
 
-  private Map<ThriftTransportKey,CachedConnections> cache = new HashMap<>();
-  private Map<ThriftTransportKey,Long> errorCount = new HashMap<>();
-  private Map<ThriftTransportKey,Long> errorTime = new HashMap<>();
-  private Set<ThriftTransportKey> serversWarnedAbout = new HashSet<>();
+  private Map<ThriftTransportKey,CachedConnections> cache = new ConcurrentHashMap<>();
+  private Map<ThriftTransportKey,Long> errorCount = new ConcurrentHashMap<>();
+  private Map<ThriftTransportKey,Long> errorTime = new ConcurrentHashMap<>();
```

Comment:
Since all access to these maps is protected by sync, would be more performant to use a HashMap.